## Importing and adjusting the dataset ##

In [68]:
import pandas as pd
import numpy as np

In [69]:
df = pd.read_csv(r"C:\Users\letru\Documents\Flatiron\Phase-3-Project\dsc-phase-3-project/Terry_stops.csv")

In [70]:
df

Subject Age Group   Subject ID   GO / SC Num  Terry Stop ID  \
0                     -           -1  2.015000e+13         101039   
1                     -           -1  2.015000e+13         101562   
2                     -           -1  2.015000e+13          54348   
3                     -           -1  2.014000e+13          92317   
4                     -           -1  2.015000e+13          28806   
...                 ...          ...           ...            ...   
47991      56 and Above  24919600373  2.021000e+13    24917463006   
47992      56 and Above  25405115029  2.021000e+13    25405364726   
47993      56 and Above  25405630254  2.021000e+13    25405654732   
47994      56 and Above  25513841558  2.021000e+13    25513816266   
47995      56 and Above  25543575280  2.021000e+13    25543626477   

      Stop Resolution Weapon Type Officer ID  Officer YOB Officer Gender  \
0       Field Contact        None       6613         1971              M   
1       Field Contact        None       7582         1966              M   
2       Field Contact        None       6081         1962              M   
3              Arrest        None       7500         1984              M   
4       Field Contact        None       5670         1965              M   
...               ...         ...        ...          ...            ...   
47991   Field Contact           -       8386         1986              M   
47992   Field Contact           -       8633         1988              M   
47993   Field Contact           -       7456         1979              M   
47994   Field Contact           -       8609         1997              M   
47995   Field Contact           -       7763         1990              M   

                    Officer Race  ... Reported Time  \
0                  Not Specified  ...      13:46:00   
1             Hispanic or Latino  ...       4:08:00   
2                          White  ...       2:47:00   
3      Black or African American  ...      11:32:00   
4                          White  ...       7:59:00   
...                          ...  ...           ...   
47991                      White  ...      12:39:30   
47992  Black or African American  ...      15:37:06   
47993                      White  ...      17:41:32   
47994          Two or More Races  ...       5:55:58   
47995                      White  ...       2:34:57   

                                     Initial Call Type Final Call Type  \
0                                                    -               -   
1                                                    -               -   
2                                                    -               -   
3                                                    -               -   
4                                                    -               -   
...                                                ...             ...   
47991  HAZ - POTENTIAL THRT TO PHYS SAFETY (NO HAZMAT)          #NAME?   
47992                 DIST - IP/JO - DV DIST - NO ASLT          #NAME?   
47993   ASLT - IP/JO - WITH OR W/O WPNS (NO SHOOTINGS)          #NAME?   
47994  HAZ - IMMINENT THRT TO PHYS SAFETY (NO HAZ MAT)          #NAME?   
47995            ASLT - IP/JO - PERSON SHOT OR SHOT AT          #NAME?   

                      Call Type                      Officer Squad  \
0                             -             WEST PCT 2ND W - QUEEN   
1                             -       SOUTH PCT 1ST W - R/S RELIEF   
2                             -              NORTH PCT 3RD W - BOY   
3                             -           SOUTH PCT 1ST W - ROBERT   
4                             -                                NaN   
...                         ...                                ...   
47991                       911      SOUTHWEST PCT 2ND W - WILLIAM   
47992                       911        WEST PCT 2ND W - K/Q RELIEF   
47993                       911      NORTH PCT 2ND W - JOHN RELIEF   
47994  TELEPHO

This data was is based in Seattle from 2015 - present.

In [71]:
df = df[['Subject Age Group', 'Subject ID', 'Stop Resolution',
         'Weapon Type', 'Officer ID', 'Officer YOB', 
         'Officer Gender', 'Officer Race', 'Subject Perceived Race', 
         'Subject Perceived Gender', 'Reported Date', 'Initial Call Type','Call Type',
         'Officer Squad', 'Arrest Flag', 'Frisk Flag',
         'Precinct', 'Sector', 'Beat' ]]

In [72]:
df.columns = ['subject_age', 'subject_id', 'stop_resolution',
         'weapon_type', 'officer_id', 'officer_age', 
         'officer_gender', 'officer_race', 'subject_race',
         'subject_gender', 'date', 'initial_call_type', 'call_type',
         'officer_squad', 'arrest', 'frisk',
         'precinct', 'sector', 'beat' ]

In [74]:
df = df.astype({'subject_age': 'category', 'officer_gender':'category', 
                'officer_race': 'category', 'subject_race':'category', 
                'subject_gender':'category', 'date':'datetime64','arrest':'category', 
                'frisk':'category','precinct':'category', 'sector':'category',
                'beat': 'category'})

In [75]:
df.replace('-', np.nan, inplace = True)

In [76]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47996 entries, 0 to 47995
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   subject_age        46433 non-null  category      
 1   subject_id         47996 non-null  int64         
 2   stop_resolution    47996 non-null  object        
 3   weapon_type        35428 non-null  object        
 4   officer_id         47972 non-null  object        
 5   officer_age        47996 non-null  int64         
 6   officer_gender     47996 non-null  category      
 7   officer_race       47996 non-null  category      
 8   subject_race       46146 non-null  category      
 9   subject_gender     47761 non-null  category      
 10  date               47996 non-null  datetime64[ns]
 11  initial_call_type  34792 non-null  object        
 12  call_type          34792 non-null  object        
 13  officer_squad      47391 non-null  object        
 14  arrest

## Exploring the data

In [77]:
#Looking at race of the subject.
df.subject_race.value_counts(ascending = False, dropna = False)

White                                        23452
Black or African American                    14304
Unknown                                       2735
NaN                                           1850
Hispanic                                      1684
Asian                                         1570
American Indian or Alaska Native              1375
Multi-Racial                                   809
Other                                          152
Native Hawaiian or Other Pacific Islander       65
Name: subject_race, dtype: int64

- Total: 46,146 races counted.
    * White subjects accounted for 51% of terry stops.
    * Black subjects accounted for 31% of terry stops.
    * Unknowns and NAN accounted for 10%.
        - It is worth mentioning here that there are ~4500 subjects of unidentified race.

In [80]:
#Looking at race of officers.
df.officer_race.value_counts(ascending = False, dropna = False)

White                            36171
Hispanic or Latino                2798
Two or More Races                 2726
Asian                             2079
Black or African American         1918
Not Specified                     1493
Nat Hawaiian/Oth Pac Islander      458
American Indian/Alaska Native      318
Unknown                             35
Name: officer_race, dtype: int64

In [88]:
#age of subject
df.subject_age.value_counts(dropna = False)

26 - 35         15984
36 - 45         10240
18 - 25          9576
46 - 55          6212
56 and Above     2439
17-Jan           1982
NaN              1563
Name: subject_age, dtype: int64

In [95]:
#age of officer
df.officer_age.value_counts(dropna = False, ascending = False).head()

31    3162
30    2957
33    2869
32    2831
29    2711
Name: officer_age, dtype: int64

In [84]:
df.officer_age.fillna(int(df.officer_age.mean()), inplace = True)

In [85]:
df['officer_age'] = (df.date.dt.year - df.officer_age)

In [90]:
df = df.drop(df[df.officer_age >70].index)

In [91]:
df.officer_age.describe()

count    47961.000000
mean        34.566648
std          8.420693
min         21.000000
25%         28.000000
50%         33.000000
75%         39.000000
max         70.000000
Name: officer_age, dtype: float64